Remove Maven client at runtime + Provide a way to load Druid extensions through local file system

guobingkun opened the pull request:
This is a follow up of https://groups.google.com/forum/#!searchin/druid-development/proposal/druid-development/7KJCsQ9GvGo/u3XMyNLmeAIJ

This pull request does the following, 
1) Remove maven client from downloading extensions at runtime
2) Provide a way to load Druid extensions and hadoop dependencies through file system.
3) Refactor pull-deps so that it can download extensions into extension directories.
4) Add documents on how to use this new extension loading mechanism

guobingkun commented:
Close and reopen to restart the build.

fjy commented on docs/content/operations/including-extensions.md:
```diff
@@ -13,22 +13,137 @@ Druid extensions can be specified in the `common.runtime.properties`. There are
+
+If you don't specify `druid.extensions.toLoad`, Druid will load all the extensions under root extension directory.
+
+### How to prepare extensions directories using `tools pull-deps`?
```
I feel this should be the default option as it is the easier option and be placed before the instructions to load manual setup.

gianm replied:
I think if we keep including all the base extensions in release tarballs (like @xvrl's newer 0.8.1 builds) then this point is not too important, since most people will be using the prebuilt extensions dir. Only people wanting to use their own extensions would have to decide how to get the jars.

fjy commented on docs/content/operations/including-extensions.md:
```diff
@@ -13,22 +13,137 @@ Druid extensions can be specified in the `common.runtime.properties`. There are
+
+If you don't want to manually setup the extension directory described above, you can use `tools pull-deps` to generate it.
+
+1) Specify `druid.extensions.extensionsDir` and `druid.extensions.hadoopDependenciesDir` for `tools pull-deps`.  After `tools pull-deps` finishes, if you go to the localtions specified in `druid.extensions.extensionsDir` and `druid.extensions.hadoopDependenciesDir`, you will see the extension directories there.
```
can we provide an example of how to run pull-deps?

fjy replied:
we might also want to update tutorials that include extensions (i.e. loading data tutorials) to use pull-deps

guobingkun replied:
Sure, will do!

guobingkun replied:
@fjy I created a new page for pull-deps with a full example and documentation, and referenced it in including-extensions page

fjy commented on docs/content/operations/including-extensions.md:
````diff
@@ -13,22 +13,137 @@ Druid extensions can be specified in the `common.runtime.properties`. There are
+
+Suppose you specify `druid.extensions.hadoopDependenciesDir=/usr/local/druid/hadoop_druid_dependencies`, and you want to prepare both `hadoop-client` 2.3.0 and 2.4.0,
+
+Then under ```hadoop_druid_dependencies```, it should look like this,
````
we'll need to update the docs around how to get Druid working with different versions of Hadoop

xvrl replied:
we also need to add docs on how to move from the old scheme to the new scheme in the release notes.

guobingkun replied:
@fjy I think http://druid.io/docs/latest/operations/other-hadoop.html still works with current changes, which part of Hadoop docs do you want me to update?

fjy replied:
What I'd like to give people an easier alternative to get their version of Hadoop working with Druid and how they can do that now that this work has completed.

guobingkun replied:
@fjy  Done updating Hadoop docs. Added section on how to include Hadoop dependencies, refined a bit the section on how to work with different versions of Hadoop.

fjy commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -171,199 +162,105 @@ static void clearLoadedModules()
+      for (final String extensionName : toLoad) {
+        final File extensionDir = new File(rootExtensionsDir, extensionName);
+        if (!extensionDir.exists()) {
+          throw new RuntimeException(
```
perhaps ISE is better here?

drcrallen replied:
IOException?

guobingkun replied:
I think ISE fits here better since there isn't really a failure in I/O operation?

drcrallen replied:
`java.io.FileNotFoundException` is IOException

guobingkun replied:
@drcrallen I've made some changes for checking directories. Do you think IOException still applies for cases that 1. Extension directory exists but is not a directory 2. Extension directory doesn't exist. I don't think it applies for case 1.

drcrallen replied:
Thinking a bit more on it, IOException should probably indicate an operation "interrupted or failed", which could include not being able to make the directory or it doesn't exist after an attempt to make it. 

In the case where the user supplied an option which is not a valid location, IllegalArgumentException (IAE) should be good?

ISE is over-used in druid at the moment, it is supposed to signal "method has been invoked at an illegal or inappropriate time".

guobingkun replied:
Yes, and also IOException is a checked exception that is supposed to be somewhat handled. Since we are going to fail directly in this case, it makes more sense to use a RuntimeException or its subclass here.

fjy commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -171,199 +162,105 @@ static void clearLoadedModules()
+      final File versionDir = new File(hadoopDependencyDir, artifact.getVersion());
+      // find the hadoop dependency with the version specified in coordinate
+      if (!hadoopDependencyDir.exists() || !versionDir.exists()) {
+        throw new RuntimeException(
```
ISE?

drcrallen replied:
IOException?

guobingkun replied:
Same above.

fjy commented:
@guobingkun I have a few minor comments but overall I am extremely excited for this to get merged. How backwards compatible is this? It seems like it is backwards compatible and we don't need to update the version.

drcrallen commented on processing/src/main/java/io/druid/guice/ExtensionsConfig.java:
```diff
@@ -33,10 +34,6 @@
 
-  @JsonProperty
-  @NotNull
-  private List<String> coordinates = ImmutableList.of();
```
Can this actually be left in, and give an error at startup if it is specified?

There doesn't seem to be a migration plan overall here. (or I could have just missed it)

guobingkun replied:
I believe it's sufficient to mention this deprecation in the release note? Even if people set this property, it will be ignored and not cause any harm.

drcrallen replied:
> not cause any harm

Not true, if they didn't realize one of the nodes had not been updated to the "new" config, then it will not launch properly, or will have missing functionality that may not even report on bootup.

guobingkun replied:
Not sure if I fully understand this. Are you worrying about people not being aware of one of the nodes hadn't been updated to the latest version? If that node fails on bootup then people will realize it's not updated properly. If that node just runs the old version silently, then failing on "specifying ```druid.extensions.coordinates```" will not help either since that node didn't pick up this code. Or did I miss something?

drcrallen replied:
If someone is updating their cluster to the latest binaries, it could be easy for them to miss the release notes that the `druid.extensions.coordinates` changed. And it can cause late failures.

Since the config change is not handled at start time, and there is no migration method to warn people of the config change (other than release notes) AND failure to migrate to the new config can cause node failure due to extensions not being present when they are expected to be present, that would be an incompatible change.

Personally, I'm ok with eliminating the setting provided that there is an informative error on startup (or even better, an informative warning and transparent handling of migration until next major version release). Would be worth getting input from the other contributors though.

drcrallen commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -171,199 +162,105 @@ static void clearLoadedModules()
+   *
+   * @return an array of druid extension files that will be loaded by druid process
+   */
+  public static File[] getExtensionFilesToLoad(ExtensionsConfig config)
```
How pedantic do we want to be here?

Should we be checking that not only these files exist but that they are directories?

guobingkun replied:
Good question :) I think we should fail loudly if extension directory is a file.

drcrallen commented on services/src/main/java/io/druid/cli/CliHadoopIndexer.java:
```diff
@@ -76,14 +75,10 @@ public void run()
-            aetherClient, coordinate, extensionsConfig.getDefaultVersion()
-        );
-        extensionURLs.addAll(Arrays.asList(((URLClassLoader) coordinateLoader).getURLs()));
+      for (final File extension : Initialization.getExtensionFilesToLoad(extensionsConfig)) {
```
NPE if extensions dir is empty

guobingkun replied:
If extensions directory is empty, it will return a zero-length array of ```File[]```. Or did I miss something?

drcrallen replied:
you are correct, I must have been thinking of something else. From java docs:
```The array will be empty if the directory is empty.  Returns {@code null} if this abstract pathname does not denote a directory, or if an I/O error occurs.```

drcrallen commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -171,199 +162,105 @@ static void clearLoadedModules()
+    final File rootExtensionsDir = new File(config.getExtensionsDir());
+    final List<String> toLoad = config.getToLoad();
+    if (toLoad == null) {
+      extensionsToLoad = rootExtensionsDir.exists() ? rootExtensionsDir.listFiles() : new File[]{};
```
Suggest putting more here.

1. There is no guarantee the rootExtensionsDir has anything in it. needs to check if listFiles is null
2. For the files that are in rootExtensionsDir, there is no guarantee that they are directories (example: zero length `.keep` file) The result of listFiles should be filtered down to only directories.

drcrallen replied:
I'm also curious now how this acts if you have symlinks. For example, if you had a master "extensions" directory, and some other extensions_historical, extensions_broker, extensions_peon kind of directories, and symlinked to the appropriate entry in the master "extensions", would that work in this workflow?

guobingkun replied:
@drcrallen I tried the symlinks you set up, and it still works.

drcrallen commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -171,199 +162,105 @@ static void clearLoadedModules()
+      final File hadoopDependencyDir = new File(rootHadoopDependenciesDir, artifact.getArtifactId());
+      final File versionDir = new File(hadoopDependencyDir, artifact.getVersion());
+      // find the hadoop dependency with the version specified in coordinate
+      if (!hadoopDependencyDir.exists() || !versionDir.exists()) {
```
Suggest also checking if they are directories

xvrl commented on docs/content/configuration/index.md:
```diff
@@ -23,7 +23,9 @@ Many of Druid's external dependencies can be plugged in as modules. Extensions c
 |`druid.extensions.remoteRepositories`|This is a JSON Array list of remote repositories to load dependencies from. If this is not set to '[]', Druid will try to download extensions at the specified remote repository.|["http://repo1.maven.org/maven2/", "https://metamx.artifactoryonline.com/metamx/pub-libs-releases-local"]|
 |`druid.extensions.localRepository`|.  The way maven gets dependencies is that it downloads them to a "local repository" on your local disk and then collects the paths to each of the jars.  This specifies the directory to consider the "local repository". If this is set, remoteRepositories is not required.|`~/.m2/repository`|
-|`druid.extensions.coordinates`|This is a JSON array of "groupId:artifactId[:version]" maven coordinates. For artifacts without version specified, Druid will append the default version. Notice: extensions explicitly specified in this property will have precedence over ones included in the classpath when Druid loads extensions. If there are duplicate extensions, Druid will only load ones explicitl ...
+|`druid.extensions.extensionsDir`|The root extension directory where user can put extensions related files. Druid will load extensions stored under this directory.|`druid_extensions` (This is a relative path to Druid's working directory)|
```
`druid.extensions.directory` would seem more natural

fjy replied:
:+1:

guobingkun replied:
Changed to ```druid.extensions.directory```

xvrl commented on docs/content/configuration/index.md:
```diff
@@ -23,7 +23,9 @@ Many of Druid's external dependencies can be plugged in as modules. Extensions c
+|`druid.extensions.extensionsDir`|The root extension directory where user can put extensions related files. Druid will load extensions stored under this directory.|`druid_extensions` (This is a relative path to Druid's working directory)|
+|`druid.extensions.hadoopDependenciesDir`|The root hadoop dependencies directory where user can put hadoop related dependencies files. Druid will load the dependencies based on the hadoop coodinate specified in the hadoop index task.|`hadoop_druid_dependencies` (This is a relative path to Druid's working directory|
+|`druid.extensions.toLoad`|A list of extensions to load from extension directories by Druid.|If not specified, Druid will load all the extensions under `druid.extensions.extensionsDir`|
 |`druid.extensions.defaultVersion`|Version to use for extension artifacts without version information.|`druid-server` artifact version.|
```
do we still need `druid.extensions.defaultVersion` ?

fjy replied:
backwards compat?

xvrl replied:
at runtime we wouldn't have a version anymore, we would just have folders with whatever version was downloaded. It seems a default version is only required for pull-deps?

xvrl replied:
Also, `druid.extensions.defaultVersion` only makes sense in combination with `druid.extensions.coordinates`, which was removed.

guobingkun replied:
@xvrl I should've made the documentation more informative. 
```druid.extensions.defaultVersion``` is used with ```tool-deps```, when you use `-c` option, you can specify a maven coordinate in the same way as we specify ```druid.extensions.coordinates```. For example ```tools pull-deps -c
io.druid.extensions:mysql-metadata-storage```, in this case ```druid.extensions.defaultVersion``` will be used to complete this maven coordinate if specified.

xvrl replied:
@guobingkun in that case, we should probably just make the default version an argument to pull-deps then, since there is no need for it at runtime anymore. It's confusing to have to specify some things (e.g. coordinates) as command line arguments and others (e.g. version) as part of the configuration.

guobingkun replied:
@xvrl Will make it an optional argument for ```pull-deps```, just like ```druid.extensions.localRepostiory```

xvrl commented on services/src/main/java/io/druid/cli/PullDependencies.java:
```diff
@@ -19,33 +19,76 @@
 @Command(
     name = "pull-deps",
-    description = "Pull down dependencies to the local repository specified by druid.extensions.localRepository"
+    description = "Pull down dependencies to the local repository specified by druid.extensions.localRepository, extensions directory specified by druid.extensions.extensionsDir and hadoop depenencies directory specified by druid.extensions.hadoopDependenciesDir"
```
do we still need localRepository, since the extensions directory replaces that function?

fjy replied:
backwards compat?

xvrl replied:
As it is this PR does not seem to preserve backwards compatibility. If we are not using aether to load at runtime anymore, then we shouldn't need the local repository. We only need to create a temporary repository for pull-deps to download jars, and then lay dependencies out into the extensions folder as needed.

guobingkun replied:
@xvrl I think local repository still serves as a cache for downloading maven coordinates, so that if you call ```tool-deps``` multiple times for the same maven coordinate, only the first time will actually need to download files from internet, which takes a bit time to finish.

xvrl replied:
@guobingkun yes, we still need some place to cache it, but I think confusing to the user to mention it as part of the pull-deps command. It might make sense to add a new optional argument to make the local cache repo configurable, and alternatively just create a temporary folder if the user doesn't specify it, that should make things easier to use.

guobingkun replied:
@xvrl Yes, making it an optional argument for ```pull-deps``` sounds good to me since it's the only place that uses it.

guobingkun replied:
@xvrl  Done making optional arguments for ```druid.extensions.localRepostiory```, ```druid.extensions.defaultVersion```. Besides those, I also removed ```druid.extensions.remoteRepositories``` and made it an optional argument.

drcrallen commented on services/src/main/java/io/druid/cli/PullDependencies.java:
```diff
@@ -53,23 +96,265 @@
+      );
+    }
+
+    PrintStream oldOut = System.out;
```
Why is this done?

guobingkun replied:
This is to temporarily copy out a reference to the current System.out and set System.out to a noop output stream. There is a comment block above,

>    DefaultTeslaAether logs a bunch of stuff to System.out, which is annoying.  We choose to disable that
    unless debug logging is turned on.  "Disabling" it, however, is kinda bass-ackwards.  We copy out a reference
    to the current System.out, and set System.out to a noop output stream.  Then after DefaultTeslaAether has pulled
    The reference we swap things back.

drcrallen replied:
In general this seems like a bad approach, but since it is how it was done previously, it is reasonable to retain the behavior until it can be addressed.

drcrallen commented on services/src/main/java/io/druid/cli/PullDependencies.java:
```diff
@@ -53,23 +96,265 @@
+   */
+  private void createExtensionDirectory(String coordinate, File atLocation)
+  {
+    if (atLocation.exists()) {
```
suggest also checking if it is a directory.

drcrallen commented on services/src/main/java/io/druid/cli/PullDependencies.java:
```diff
@@ -53,23 +96,265 @@
+  private void createRootExtensionsDirectory(File atLocation)
+  {
+    if (!atLocation.mkdirs()) {
+      throw new RuntimeException(
```
IOException?

guobingkun replied:
For this one, I think it still makes more sense to throw a RuntimeException or its subclass IllegalArgumentException as you suggested.

guobingkun commented:
@fjy  It should be backward compatible. The only difference is that at runtime, Druid loads extensions and hadoop dependencies from local file system instead of internet.

xvrl commented on docs/content/configuration/index.md:
```diff
@@ -23,7 +23,9 @@ Many of Druid's external dependencies can be plugged in as modules. Extensions c
-|`druid.extensions.coordinates`|This is a JSON array of "groupId:artifactId[:version]" maven coordinates. For artifacts without version specified, Druid will append the default version. Notice: extensions explicitly specified in this property will have precedence over ones included in the classpath when Druid loads extensions. If there are duplicate extensions, Druid will only load ones explicitl ...
+|`druid.extensions.extensionsDir`|The root extension directory where user can put extensions related files. Druid will load extensions stored under this directory.|`druid_extensions` (This is a relative path to Druid's working directory)|
+|`druid.extensions.hadoopDependenciesDir`|The root hadoop dependencies directory where user can put hadoop related dependencies files. Druid will load the dependencies based on the hadoop coodinate specified in the hadoop index task.|`hadoop_druid_dependencies` (This is a relative path to Druid's working directory|
+|`druid.extensions.toLoad`|A list of extensions to load from extension directories by Druid.|If not specified, Druid will load all the extensions under `druid.extensions.extensionsDir`|
```
minor nit, using a verb prefixed with "to" seems a bit odd, maybe `druid.extensions.enable` sounds better?

xvrl replied:
also, it's unclear what format to specify the list in, is it a comma-separated list of directories or a json array?

guobingkun replied:
Done. Changed to ```druid.extensions.enable``` and updates its description to ```A JSON array of extensions to load from extension directories by Druid.|If not specified, Druid will load all the extensions under `druid.extensions.directory````

guobingkun commented:
Finished addressing comments. @fjy @xvrl @drcrallen @himanshug  @gianm @nishantmonu51

fjy commented:
:+1:

drcrallen commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -143,15 +120,29 @@ static void clearLoadedModules()
-      try {
-        URLClassLoader loader = getClassLoaderForCoordinates(aether, coordinate, config.getDefaultVersion());
+    if (config.searchCurrentClassloader()) {
+      for (T module : ServiceLoader.load(clazz, Initialization.class.getClassLoader())) {
```
(should not block this PR) I was looking at this part of the code recently and wondering if this should be using the thread context class loader by default instead of the Initialization.class's classloader.

guobingkun replied:
Changed to thread context class loader.

drcrallen commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -171,199 +162,111 @@ static void clearLoadedModules()
-
-        for (URL url : urls) {
-          log.info("Added URL[%s]", url);
+    final File rootExtensionsDir = new File(config.getDirectory());
```
I think the workflow lost the "check if it exists" path.

guobingkun replied:
The workflow doesn't explicitly check if it exists. if ```rootExtensionsDir``` doesn't exist, then ```listFiles()``` returns null, the return statement in this function handles that case.

drcrallen commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -171,199 +162,111 @@ static void clearLoadedModules()
-      catch (URISyntaxException e) {
-        throw Throwables.propagate(e);
-      }
+    final File rootHadoopDependenciesDir = new File(extensionsConfig.getHadoopDependenciesDir());
```
Same here about checking if exists at all.

guobingkun replied:
Similar here, it is not explicitly checked, but if ```rootHadoopDependenciesDir``` doesn't exist, and ```hadoopCoordinates``` or ```hadoopDependency``` is not empty, ISE is thrown.

drcrallen replied:
Can you check that in a unit test?

guobingkun replied:
Will do.

drcrallen commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -171,199 +162,111 @@ static void clearLoadedModules()
+   */
+  public static URLClassLoader getClassLoaderForExtension(File extension) throws MalformedURLException
+  {
+    URLClassLoader loader = loadersMap.get(extension.getName());
```
Do we actually need to keep this static map?

guobingkun replied:
Yeah, this is called from ```getFromExtensions(xxx)```, which is called twice, once in https://github.com/druid-io/druid/blob/master/services/src/main/java/io/druid/cli/Main.java#L76
and other in https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/initialization/Initialization.java#L408

drcrallen commented on services/src/main/java/io/druid/cli/PullDependencies.java:
```diff
@@ -18,34 +18,103 @@
+  @Option(
+      name = {"-c", "--coordinate"},
+      title = "coordinate",
+      description = "Extension coordinate to pull down, followed by a maven coordinate, e.g. io.druid.extensions:mysql-metadata-storage:0.8.0-rc1",
```
Probably want this to be something that can actually be found in maven central

guobingkun replied:
Done. Changed to 0.8.1-rc2

fjy replied:
err, can we only do stables or omit it altogether i.e. just write <version> instead of an actual version

guobingkun replied:
Done.

drcrallen commented on docs/content/configuration/index.md:
```diff
@@ -21,10 +21,9 @@ Many of Druid's external dependencies can be plugged in as modules. Extensions c
-|`druid.extensions.defaultVersion`|Version to use for extension artifacts without version information.|`druid-server` artifact version.|
+|`druid.extensions.directory`|The root extension directory where user can put extensions related files. Druid will load extensions stored under this directory.|`druid_extensions` (This is a relative path to Druid's working directory)|
+|`druid.extensions.hadoopDependenciesDir`|The root hadoop dependencies directory where user can put hadoop related dependencies files. Druid will load the dependencies based on the hadoop coodinate specified in the hadoop index task.|`hadoop_druid_dependencies` (This is a relative path to Druid's working directory|
+|`druid.extensions.enable`|A JSON array of extensions to load from extension directories by Druid.|If not specified, Druid will load all the extensions under `druid.extensions.directory`|
```
Suggest making this `druid.extensions.toLoad` or similar. having "enable" makes it sound like a boolean.

guobingkun replied:
I am fine with either. I took @xvrl 's suggestion at https://github.com/druid-io/druid/pull/1638#discussion_r37678389

nishantmonu51 replied:
+1 on renaming the property to something else than enable. probably druid.extensions.list ?

guobingkun replied:
After thinking a bit more on it, I've decided to rename it to ```druid.extensions.load``` as it indicates the values on the right side will be loaded, and doesn't sound like a boolean nor prefix with "to".

guobingkun replied:
oops, I meant to type ```druid.extensions.loadList```, updated the comment.

xvrl commented on docs/content/misc/tasks.md:
```diff
@@ -139,7 +139,8 @@ The Hadoop Index Task is used to index larger data sets that require the paralle
 |spec|A Hadoop Index Spec. See [Batch Ingestion](../ingestion/batch-ingestion.html)|yes|
 |hadoopCoordinates|The Maven \<groupId\>:\<artifactId\>:\<version\> of Hadoop to use. The default is "org.apache.hadoop:hadoop-client:2.3.0".|no|
-
+|hadoopDependencyCoordinates|A JSON array of Hadoop dependency coordinates that will override the default Hadoop coordinates and `hadoopCoordinates`.|no|
```
it's unclear from the documentation what the difference between hadoopCoordinates and hadoopDependencyCoordinates is. If one overrides the other, why do we have two?

guobingkun replied:
The main difference is that ```hadoopCoordinates``` is a String. I think @gianm could answer this question better?

gianm replied:
hadoopDependencyCoordinates is a list and is the preferred one, hadoopCoordinates is a string and is for backwards compatibility. IMO we should remove hadoopCoordinates from the docs.

fjy replied:
:+1:

guobingkun replied:
Removed hadoopCoordinates from doc.

xvrl commented on docs/content/operations/pull-deps.md:
```diff
@@ -0,0 +1,90 @@
+
+`pull-deps` has several command line options, they are as follows:
+
+`-c` or `--coordinate`     
```
can we clarify that this option can be specified multiple times?

guobingkun replied:
Updated doc.

xvrl commented on docs/content/operations/pull-deps.md:
```diff
@@ -0,0 +1,90 @@
+
+    Don't pull down the default hadoop coordinate, i.e., org.apache.hadoop:hadoop-client:2.3.0. If `-h` option is supplied, then default hadoop coordinate will not be downloaded.
+
+`-clean`
```
can we prefix options that have more than one letter with `--` i.e. `--clean` to be consistent?

guobingkun replied:
Done.

xvrl commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -132,9 +111,7 @@ static void clearLoadedModules()
-   * Look for extension modules for the given class from both classpath and druid.extensions.coordinates.
-   * Extensions explicitly specified in druid.extensions.coordinates will be loaded first, if there is a duplicate
-   * extension from classpath, it will be ignored.
+   * Look for extension modules for the given class from both classpath and extensions directory.
```
can we clarify which one takes precedence? will it prefer the extension found on the classpath or the one from the directory?

guobingkun replied:
Updated comments. Classpath will take precedence since ```ServiceLoader.load(clazz, loader)``` will load extensions specified in classpath even when we only want it to load extensions in the extension directory.

nishantmonu51 commented on docs/content/operations/pull-deps.md:
````diff
@@ -0,0 +1,90 @@
+
+Suppose you want to download ```druid-examples```, ```mysql-metadata-storage``` and ```hadoop-client```(both 2.3.0 and 2.4.0) with a specific version, you can run `pull-deps` command with `-c io.druid.extensions:druid-examples:0.8.0-rc1`, `-c io.druid.extensions:mysql-metadata-storage:0.8.0-rc1`, `-h org.apache.hadoop:hadoop-client:2.3.0` and `-h org.apache.hadoop:hadoop-client:2.4.0`, an example ...
+
+```java -classpath "/my/druid/library/*" io.druid.cli.Main tools pull-deps -clean -c io.druid.extensions:mysql-metadata-storage:0.8.0-rc1 -c io.druid.extensions:druid-examples:0.8.0-rc1 -h org.apache.hadoop:hadoop-client:2.3.0 -h org.apache.hadoop:hadoop-client:2.4.0```
````
Is it required/optional to add maven version for the extensions to download ?

guobingkun replied:
default maven version is optional for extension coordinate. I added more docs on this.

nishantmonu51 commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -171,199 +162,111 @@ static void clearLoadedModules()
+  /**
+   * Find all the extension files that should be loaded by druid.
+   * <p/>
+   * If user explicitly specifies druid.extension.toLoad, then it will look for those extensions under root
```
make this property name consistent with the docs.

guobingkun replied:
Done.

guobingkun commented:
Done second scan of review comments. Added unit tests that should've covered most workflow.
@fjy @xvrl @drcrallen @himanshug @gianm @nishantmonu51 @cheddar

cheddar commented:
@guobingkun let's make sure this doesn't break the tarball.  Then this is probably merge-able for 0.8.2.

cheddar commented:
It loos like we will also need to bundle the mysql connector into a separate tarball, not the main release one.

guobingkun commented:
@cheddar Sure, will do!

xvrl commented:
@cheddar @guobingkun given that we are removing existing configuration options, and are not fully backwards compatible, i.e. changing the way most users have to deploy Druid, should we bump version up to 0.9 after merging this?

guobingkun commented:
I  made some changes to make sure the tarball still works,
1. Added an exec plugin so that ```mvn clean install``` will generate all the extensions directories, there will be a ```druid_extensions``` directory that contains all the extensions (except ```mysql-metadata-storage```) in the tarball.
2. Added another assembly step so that ```mysql-metadata-storage``` will be packaged in a separate tarball. User can go into ```druid_extensions``` and untar ```mysql-metadata-storage--bin.tar.gz``` there.
3. Modified common runtime properties so that the extension load list still contains ```druid-examples```,```druid-kafka-eight```
4. Removed extension-distribution since it is replaced by extension directoies.
5. Updated tutorial and firewall docs.
@fjy @xvrl @gianm @drcrallen @nishantmonu51 @himanshug @cheddar

guobingkun commented:
Forgot to mention I also updated Build From Source doc. Now ```mvn clean install``` generates the tarballs.

fjy commented on docs/content/development/build.md:
```diff
@@ -16,11 +16,14 @@ To do so, run these commands:
-`services/target/druid-VERSION-bin.tar.gz`.
+`distribution/target/druid-VERSION-bin.tar.gz`.
+
+This will also create a tarball that contains `mysql-metadata-storage` extension under
```
we will need to also include this information on the page about metadata stores

guobingkun replied:
added into metadata-storage.md

fjy commented on docs/content/operations/other-hadoop.md:
```diff
@@ -1,20 +1,73 @@
 ---
+# Work with different versions of Hadoop
+
 Working with different versions of Hadoop may require a bit of extra work for the time being. We will make changes to support different Hadoop versions in the near future. If you have problems outside of these instructions, please feel free to contact us in IRC or on the [forum](https://groups.google.com/forum/#!forum/druid-development).
```
we can probably delete this paragraph

guobingkun replied:
done.

fjy commented on docs/content/operations/other-hadoop.md:
```diff
@@ -1,20 +1,73 @@
+
+To make this work, follow the steps below
+
+(1) Specify `druid.extensions.hadoopDependenciesDir` (root directory for Hadoop related dependencies). If you don' specify it, Druid will use its default value, see [Configuration](../configuration/index.html).
```
don't*

guobingkun replied:
fixed.

fjy commented on docs/content/operations/pull-deps.md:
```diff
@@ -0,0 +1,94 @@
+
+`-d` or `--defaultVersion`
+
+    Version to use for extension coordinate that doesn't have a version information. For example, if extension coordiante is `io.druid.extensions:mysql-metadata-storage`, and default version is `0.8.0`, then this coordinate will be treated as `io.druid.extensions:mysql-metadata-storage:0.8.0`
```
coordinate*

guobingkun replied:
fixed typo

fjy commented on docs/content/toc.textile:
```diff
@@ -64,6 +64,7 @@ h2. Operations
 * "Different Hadoop Versions":../operations/other-hadoop.html
 * "Multitenancy Considerations":../operations/multitenancy.html
 * "Performance FAQ":../operations/performance-faq.html
+* "Pull Dependencies":../operations/pull-deps.html
```
I don't believe this deserves a TOC entry and the file should be linked via the "Including Extensions" page

guobingkun replied:
agree

fjy commented on docs/content/tutorials/tutorial-a-first-look-at-druid.md:
```diff
@@ -74,6 +74,8 @@ You should see a bunch of files:
 * run_example_server.sh
 * run_example_client.sh
 * LICENSE, config, examples, lib directories
+* druid_extenions (This folder contains all the extensions that could be loaded by Druid. Note that extension `mysql-metadata-storage` is packaged in a separate tarball that can be downloaded from [here](http://druid.io/downloads.html). See [Including Extensions](../operations/including-extensions.html) for more information about loading extensions.
```
druid_extensions*

guobingkun replied:
fixed typo.

fjy commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -132,9 +111,9 @@ static void clearLoadedModules()
-   * Extensions explicitly specified in druid.extensions.coordinates will be loaded first, if there is a duplicate
-   * extension from classpath, it will be ignored.
+   * Look for extension modules for the given class from both classpath and extensions directory. A user should never
+   * put two same extensions in classpath and extensinos directory, if he/she does that, the one that is in the
```
extensions*

fjy replied:
"the same two extensions" not "put two same..."

guobingkun replied:
fixed.

xvrl commented on docs/content/configuration/index.md:
```diff
@@ -21,10 +21,9 @@ Many of Druid's external dependencies can be plugged in as modules. Extensions c
-|`druid.extensions.coordinates`|This is a JSON array of "groupId:artifactId[:version]" maven coordinates. For artifacts without version specified, Druid will append the default version. Notice: extensions explicitly specified in this property will have precedence over ones included in the classpath when Druid loads extensions. If there are duplicate extensions, Druid will only load ones explicitl ...
-|`druid.extensions.defaultVersion`|Version to use for extension artifacts without version information.|`druid-server` artifact version.|
+|`druid.extensions.directory`|The root extension directory where user can put extensions related files. Druid will load extensions stored under this directory.|`druid_extensions` (This is a relative path to Druid's working directory)|
+|`druid.extensions.hadoopDependenciesDir`|The root hadoop dependencies directory where user can put hadoop related dependencies files. Druid will load the dependencies based on the hadoop coodinate specified in the hadoop index task.|`hadoop_druid_dependencies` (This is a relative path to Druid's working directory|
```
typo, missing r in coordinate, we should also name the field in the index task explicitly or link to it.

guobingkun replied:
Fixed the typo + mention ```druid.extensions.hadoopDependenciesDir``` in hadoop index task doc.

xvrl commented on docs/content/configuration/index.md:
```diff
@@ -21,10 +21,9 @@ Many of Druid's external dependencies can be plugged in as modules. Extensions c
-|`druid.extensions.defaultVersion`|Version to use for extension artifacts without version information.|`druid-server` artifact version.|
+|`druid.extensions.directory`|The root extension directory where user can put extensions related files. Druid will load extensions stored under this directory.|`druid_extensions` (This is a relative path to Druid's working directory)|
+|`druid.extensions.hadoopDependenciesDir`|The root hadoop dependencies directory where user can put hadoop related dependencies files. Druid will load the dependencies based on the hadoop coodinate specified in the hadoop index task.|`hadoop_druid_dependencies` (This is a relative path to Druid's working directory|
+|`druid.extensions.loadList`|A JSON array of extensions to load from extension directories by Druid.|If not specified, Druid will load all the extensions under `druid.extensions.directory`|
```
we should explicitly define the default, is it `null` or empty list `[]` ? does empty list make it load no extensions at all or does it cause it to revert to loading everything? We should also move the behavior in the description field, and not put it as part of the default field.

guobingkun replied:
Fixed. The default value is null if ```druid.extensions.loadList``` is not specified, and it will make Druid load all extensions under extension directory. If it is empty list ```[]```, no extensions will be loaded.

fjy commented:
:+1: But before merging we should decide on the new version of Druid and mysql packaging

xvrl commented on docs/content/tutorials/firewall.md:
```diff
@@ -4,23 +4,7 @@ layout: doc_page
-
-5. Once all dependencies have been downloaded successfully, replicate the `repo` directory to the machine behind the firewall; e.g. `/opt/druid-<version>/repo`
-6. Create property `druid.extensions.localRepository=`*`path to repo directory`* in the *`Druid Directory`*`/config/_common/common.runtime.properties` file; e.g. `druid.extensions.localRepository=/opt/druid-<version>/repo`
+When you are behind a firewall, the IRC wikipedia channels that feed realtime data into Druid will not be accessible. To workaround this challenge, you will need to make the Wikipedia example GeoLite DB dependency available offline, see below.
```
This sentence is a bit confusing, making GeoLite avaible offline will not solve the IRC problem as the sentence implies.
There are two problems, one is that IRC may be firewalled, in which case there's nothing you can do. The other problem is that downloading data from maxmind may not be possible, in which case GeoLite needs to be downloaded offline.

guobingkun replied:
updated the doc so that it reflects your comment.

xvrl commented on docs/content/tutorials/tutorial-a-first-look-at-druid.md:
```diff
@@ -74,6 +74,8 @@ You should see a bunch of files:
 * run_example_server.sh
 * run_example_client.sh
 * LICENSE, config, examples, lib directories
+* druid_extensions (This folder contains all the extensions that could be loaded by Druid. Note that extension `mysql-metadata-storage` is packaged in a separate tarball that can be downloaded from [here](http://druid.io/downloads.html). See [Including Extensions](../operations/including-extensions.html) for more information about loading extensions.
```
Did we adjust the examples to make sure we didn't load all the extensions by default? It might cause some weirdness if multiple metadata stores or other incompatible extensions get loaded at the same time.

guobingkun replied:
Yes. I modified common.runtime.properties under config/_common so that it loads druid-examples and druid-kafka-eight by default.

xvrl commented:
@cheddar unless we relicense the mysql metadata store code under GPL, I don't see how packaging it as a separate tarball would make a difference?

xvrl commented:
Assuming this has been tested, other than my last few comments about docs, and maybe resolve the mysql tabarll issue outside of this PR, I'm :+1:

drcrallen commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -143,25 +122,39 @@ static void clearLoadedModules()
-      try {
-        URLClassLoader loader = getClassLoaderForCoordinates(aether, coordinate, config.getDefaultVersion());
+    if (config.searchCurrentClassloader()) {
+      for (T module : ServiceLoader.load(clazz, Thread.currentThread().getContextClassLoader())) {
```
ServiceLoader.load checks context classloader automatically, but it probably doesn't hurt to have it here.

drcrallen commented:
This is looking good! 

As a note, it is blocked until https://groups.google.com/d/msg/druid-development/KXpLwUisJcU/eCM3GfgPBAAJ is resolved

drcrallen commented:
@guobingkun I think this is ready to merge, can you please cleanup the merge conflicts?

guobingkun commented:
@drcrallen Conflicts resolved.

se7entyse7en commented on distribution/pom.xml:
```diff
@@ -67,6 +112,20 @@
+                            <goal>single</goal>
+                        </goals>
+                        <configuration>
+                            <finalName>mysql-metdata-storage</finalName>
```
Is this a typo?

guobingkun replied:
@se7entyse7en Thanks! That's definitely a typo, I will send a PR to fix it.

drcrallen commented:
This is weird, https://github.com/druid-io/druid/pull/2067 but this still passed travis
